### ISSUE TITLE
Terrible performance from repeated non-incremental calls to static web assets targets

### ISSUE THREAD

**Issue by kzu** [dotnet/aspnetcore#49097]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Out of the box Fluent UI Blazor Server project template, including icons and emojis, results in each build taking several seconds and a performance summary showcases multiple invocations to static web assets-related targets that (upon inspection) are non-incremental (have no inputs-outputs).

![image](https://github.com/dotnet/aspnetcore/assets/169707/4d396daa-e7f8-4414-bcac-78216101952b)

(subsequent builds with no changes show same targets and mostly same times every time)

### Expected Behavior

A second build with no changes to included assets, should not cause the same targets to be called multiple times, especially for something included out of the box (I'm assuming Fluent UI is sort of official Microsoft thing).

### Steps To Reproduce

https://github.com/kzu/FluentUIBlazorServer

1. `dotnet new fluentuiblazorserver`
2. set `PublishFluentIconAssets=true` and `PublishFluentEmojiAssets=true` in the project
3. build twice with `dotnet build -clp:PerformanceSummary`

### .NET Version

8.0.100-preview.5.23303.2

### Anything else?

![image](https://github.com/dotnet/aspnetcore/assets/169707/7d0c271c-e4e1-40fb-8762-8e0763a59635)

Pretty much on every build, all those top-slowest tasks are re-run.

**Comment by vnbaaij** [comment]
>  (I'm assuming Fluent UI is sort of official Microsoft thing).

Yes,...but it is not part of the aspnetcore org. We hang out in [fluentui-blazor](https://github.com/microsoft/fluentui-blazor)

Anyways, we are changing all of this for the next version on the library. Icons and Emoji will no longer be included as static assets (but will come from separate (NuGet) packages) and the steps you mention will be removed from the project file.

Just saw a rebuild of the whole project being done in under 12 seconds and a sequential build was done in just under a second...

**Comment by javiercn** [comment]
@kzu This is an issue caused by the library.

It's adding ~19K assets to the build as content. There are some potential fixes/improvements that could be done to the incrementality of the build, however, we would need to see more packages causing this type of performance issue. In general, this scale is not something that people run into, and hence is not something we optimize for.

I believe the majority of those assets are not used by the app, and that @vnbaaij is fixing this in an upcoming release.

**Comment by kzu** [comment]
Some rather obvious things you can do to workaround this are based on the same approach taken by the GenerateAssemblyInfo in the .NET SDK, for example:

```xml
  <Import Project="$(StaticWebAssetsCacheFile)" Condition="Exists('$(StaticWebAssetsCacheFile)')"/>

  <Target Name="WriteStaticWebAssetsCacheFile" DependsOnTargets="CreateStaticWebAssetsInputsCacheFile"
          Inputs="$(StaticWebAssetsInputsCacheFile)" Outputs="$(StaticWebAssetsCacheFile)">

    <!-- Force dynamic evaluation of outdated prop so we don't remove items unless fully run -->
    <Exec Command="echo 'true'" ConsoleToMSBuild="true">
      <Output TaskParameter="ConsoleOutput" PropertyName="_OutdatedStaticWebAssetsCache" />
    </Exec>

    <ItemGroup Condition="$(_OutdatedStaticWebAssetsCache) == 'true'">
      <!-- We may have imported from items already, so clean before update -->
      <StaticWebAsset Remove="@(_CachedStaticWebAsset)"/>
    </ItemGroup>

    <UpdatePackageStaticWebAssets Assets="@(StaticWebAsset)">
      <Output TaskParameter="UpdatedAssets" ItemName="_CachedStaticWebAsset" />
    </UpdatePackageStaticWebAssets>

    <WriteItemsToFile Items="@(_CachedStaticWebAsset)" File="$(StaticWebAssetsCacheFile)" ItemName="_CachedStaticWebAsset" Overwrite="true" />

    <ItemGroup>
      <FileWrites Include="$(StaticWebAssetsCacheFile)" />
    </ItemGroup>
  </Target>

  <Target Name="ReadStaticWebAssetsCacheFile" DependsOnTargets="WriteStaticWebAssetsCacheFile">
    <!-- If we have cached items already, no need to re-read -->
    <ReadItemsFromFile File="$(StaticWebAssetsCacheFile)" Condition="@(_CachedStaticWebAsset -> Count()) == '0'">
      <Output TaskParameter="Items" ItemName="_CachedStaticWebAsset" />
    </ReadItemsFromFile>

    <ItemGroup>
      <StaticWebAsset Remove="@(StaticWebAsset)"/>
      <StaticWebAsset Include="@(_CachedStaticWebAsset)" />
    </ItemGroup>
  </Target>

  <Target Name="UpdateExistingPackageStaticWebAssets" 
          DependsOnTargets="CreateStaticWebAssetsInputsCacheFile;WriteStaticWebAssetsCacheFile;ReadStaticWebAssetsCacheFile" />
```

That alone caused the updating to happen only once. (ReadItemsFromFile and WriteItemsToFile are tasks that just dump the items and their metadata in MSBuild-formatted files that can be included wholesale afterwards in a subsequent incremental build).

Anyway, I don't think any framework-level feature that allows open-ended extensibility should ignore the possibility that component authors will use nuget packages to provide massive amounts of assets that need to be incrementally processed. Incrementality should always be the default when doing any sort of MSBuild work, IMO. 

For example, the same library offers an easily consumable Monaco editor component. That would bring another large number of assets too. It's not entirely author-friendly to just say it's their problem to figure out, if the SDK can be smarter about these things.

Just my 2cents :)
